### ISSUE TITLE
关于SWA的实现

### ISSUE THREAD

**Issue by RealDuxy** [QwenLM/Qwen2#563]
感谢千问团队的辛苦付出！
我比较关注qwen2模型所支持SWA技术，想请教一下你们在qwen2中的sliding window attention的实现。

据我的了解, 实现SWA需要
1. attention_mask 需要做划窗的mask。这一点，在你们的eager/sdpa/flash-attn中都通过成熟的方式实现了。
eager: _prepare_4d_causal_attention_mask
sdpa: _prepare_4d_causal_attention_mask_for_sdpa
flash-attn: flash_attn_func内部貌似自动实现了casual attention mask

  2. 在window内，滑动后，pask_key_value需要重新计算（当然，这非常耗时），因为key / value cache是经过apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)，携带了位置信息，滑动窗口后由于position从1-10 -> 5- 15，位置信息发生了改变。

一些改进工作，例如streamingllm对比的也是swa with re-compute

但是查看通义千问的modeling_qwen2.py中的关于past_key_value的实现，发现其似乎是直接截取缓存内的 最后window_size个token，而没有做任何缓存的重新计算。

```python
past_key = past_key[:, :, slicing_tokens:, :].contiguous()
past_value = past_value[:, :, slicing_tokens:, :].contiguous()
```

请问出于什么考虑选择这样的方式进行Cache更新呢？
我的想法是：由于RoPE + ntk-aware的引入，只要相对位置大小不变，position的绝对大小并不会特别影响结果，因此可以不重新计算key value cache, 从而减少计算量。

希望得到解答。

**Comment by jklj077** [comment]
Hi,

Most LLMs use relative positional embedding now and I don't think the recomputation is a result of the change in positional embeddings, since they don't change at all.

The thing is that the sliding window attention (SWA) implementation in Qwen2 models is similar to the one proposed in Mistral and is different from the SWA implementation in the StreamingLLM paper, where it is just called window attention.

Suppose a model is trained with a context of 2 and we're working on a sequence of length 3, [x1, x2, x3].
- For SWA in Qwen2, although x3 attends only to [x2, x3], x2 contains information from both x2 and x1, as the past key and past value is reused. There will be a mismatch between training and inference and suboptimal results are expected, which can be mitgated by training in the same way.
- For SWA in StreamingLLM, x3 attends only to [x2, x3] and x2 contains information only from x2, as the past key value is recomputed. It is strictly aligned with how the model is pretrained but it can be very slow as it effectively disables the KV cache optimization.
